ui/src/app/machines/views/MachineList/MachineList.test.js

```diff
@@ -449,4 +449,88 @@ describe("MachineList", () => {
         .props().checked
     ).toBe(true);
   });
+
+  it("can select all machines in a group if at least one is unselected", () => {
+    const state = { ...initialState };
+    state.machine.items[1].status_code = nodeStatus.DEPLOYED;
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <MemoryRouter
+          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
+        >
+          <MachineList />
+        </MemoryRouter>
+      </Provider>
+    );
+    // Select a single machine
+    wrapper
+      .find("[data-test='name-column'] input")
+      .at(0)
+      .simulate("change", {
+        target: { name: state.machine.items[0].system_id }
+      });
+    expect(
+      wrapper.find("[data-test='name-column'] input[checked=true]").length
+    ).toBe(1);
+    expect(
+      wrapper.find("[data-test='group-cell'] input[checked=true]").length
+    ).toBe(0);
+    // Check the group checkbox
+    wrapper
+      .find("[data-test='group-cell'] input")
+      .at(0)
+      .simulate("change");
+    expect(
+      wrapper.find("[data-test='name-column'] input[checked=true]").length
+    ).toBe(2);
+    expect(
+      wrapper.find("[data-test='group-cell'] input[checked=true]").length
+    ).toBe(1);
+  });
+
+  it("unselects all machines in a group if all are selected", () => {
+    const state = { ...initialState };
+    state.machine.items[1].status_code = nodeStatus.DEPLOYED;
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <MemoryRouter
+          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
+        >
+          <MachineList />
+        </MemoryRouter>
+      </Provider>
+    );
+    // Select both machines
+    wrapper
+      .find("[data-test='name-column'] input")
+      .at(0)
+      .simulate("change", {
+        target: { name: state.machine.items[0].system_id }
+      });
+    wrapper
+      .find("[data-test='name-column'] input")
+      .at(1)
+      .simulate("change", {
+        target: { name: state.machine.items[1].system_id }
+      });
+    expect(
+      wrapper.find("[data-test='name-column'] input[checked=true]").length
+    ).toBe(2);
+    expect(
+      wrapper.find("[data-test='group-cell'] input[checked=true]").length
+    ).toBe(1);
+    // Check the group checkbox
+    wrapper
+      .find("[data-test='group-cell'] input")
+      .at(0)
+      .simulate("change");
+    expect(
+      wrapper.find("[data-test='name-column'] input[checked=true]").length
+    ).toBe(0);
+    expect(
+      wrapper.find("[data-test='group-cell'] input[checked=true]").length
+    ).toBe(0);
+  });
```

Comment:
Could possibly add a test here that the group checkbox is unchecked if you check all then uncheck one.

Reply:
Thanks @huwshimi, added another test